platform/ui/src/components/ContextMenu/ContextMenuController.tsx

```diff
@@ -0,0 +1,218 @@
+import * as ContextMenuItemsBuilder from './ContextMenuItemsBuilder';
+import ContextMenu from './ContextMenu';
+
+type Obj = Record<string, unknown>;
+
+const getDefaultPosition = () => {
+  return {
+    x: 0,
+    y: 0,
+  };
+};
+
+const _getEventDefaultPosition = eventDetail => ({
+  x: eventDetail && eventDetail.currentPoints.client[0],
+  y: eventDetail && eventDetail.currentPoints.client[1],
+});
+
+const _getViewerElementDefaultPosition = viewerElement => {
+  if (viewerElement) {
+    const boundingClientRect = viewerElement.getBoundingClientRect();
+    return {
+      x: boundingClientRect.x,
+      y: boundingClientRect.y,
+    };
+  }
+
+  return {
+    x: undefined,
+    y: undefined,
+  };
+};
+
+const _getCanvasPointsPosition = (points = [], viewerElementOfReference) => {
+  const viewerPos = _getViewerElementDefaultPosition(viewerElementOfReference);
+
+  for (let pointIndex = 0; pointIndex < points.length; pointIndex++) {
+    const point = {
+      x: points[pointIndex][0] || points[pointIndex]['x'],
+      y: points[pointIndex][1] || points[pointIndex]['y'],
+    };
+    if (_isValidPosition(point) && _isValidPosition(viewerPos)) {
+      return {
+        x: point.x + viewerPos.x,
+        y: point.y + viewerPos.y,
+      };
+    }
+  }
+};
+
+const _isValidPosition = source => {
+  return source && typeof source.x === 'number' && typeof source.y === 'number';
+};
+/**
+ * Returns the context menu default position. It look for the positions of: canvasPoints (got from selected), event that triggers it, current viewport element
+ */
+const _getDefaultPosition = (canvasPoints, eventDetail, viewerElement) => {
+  function* getPositionIterator() {
+    yield _getCanvasPointsPosition(canvasPoints, viewerElement);
+    yield _getEventDefaultPosition(eventDetail);
+    yield _getViewerElementDefaultPosition(viewerElement);
+    yield getDefaultPosition();
+  }
+
+  const positionIterator = getPositionIterator();
+
+  let current = positionIterator.next();
+  let position = current.value;
+
+  while (!current.done) {
+    position = current.value;
+
+    if (_isValidPosition(position)) {
+      positionIterator.return();
+    }
+    current = positionIterator.next();
+  }
+
+  return position;
+};
+
+/**
+ * The context menu controller is a helper class that knows how
+ * to manage context menus based on the UI Customization Service.
+ * There are a few parts to this:
+ *    1. Basic controls to manage displaying and hiding context menus
+ *    2. Menu selection services, which use the UI customization service
+ *       to choose which menu to display
+ *    3. Menu item adapter services to convert menu items into displayable and actionable items.
+ */
+export default class ContextMenuController {
```

Comment:
ideally this should be the first thing in this file, and all the private `_methods` be below for readability. One should not need to scroll to the middle of the file to find the docstring